app/src/main/java/com/android/sample/model/recipe/networkData/FirestoreRecipesRepository.kt

```diff
@@ -279,8 +286,81 @@ class FirestoreRecipesRepository(private val db: FirebaseFirestore) : RecipeNetw
         .addOnFailureListener { e -> onFailure(e) }
   }
 
-  /** Will be deleted in next PR */
-  override fun listCategories(onSuccess: (List<String>) -> Unit, onFailure: (Exception) -> Unit) {
-    throw NotImplementedError("Not implemented")
+  /**
```

Comment:
LGTM!

1. **Dynamic Filtering:**  
   Great use of `let` and `takeIf` for building filters dynamically—clean and readable.

2. **Validation:**  
   Nice touch with `require` to validate `limit` upfront.

3. **Randomness:**  
   Shuffling and limiting results adds a nice layer of unpredictability.

4. **Improvement:**  
   Consider breaking the filter logic into helper methods for better readability.

5. **Documentation:**  
   Clear and helpful comments—makes the code easy to follow.

Reply:
Thanks! 😄